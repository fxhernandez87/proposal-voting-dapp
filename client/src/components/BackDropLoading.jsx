/* eslint-disable no-nested-ternary,react/destructuring-assignment */
import React, { useContext, useState, useEffect } from 'react';
import LoadingOverlay from 'react-loading-overlay';
import VotingContext from "../contexts/votingProposal";

export default ({ active, children, initializing, error }) => {

  const context = useContext(VotingContext);
  const [isActive, setIsActive] = useState(true);

  useEffect(() => {
    setIsActive(active || !!error || !!context.error);
  }, [active, error, context.error]);

  const content = context.error ? context.error.message : initializing ? "Initializing App" : "Fetching User";
  const handleClick = () => {
    setIsActive(false);
    if (context.error) {
      context.setError(null);
    }
  };
  return (
    <LoadingOverlay
      active={isActive}
      spinner={!error && !context.error}
      text={content}
      onClick={handleClick}
    >
      {children}
    </LoadingOverlay>
  );
}