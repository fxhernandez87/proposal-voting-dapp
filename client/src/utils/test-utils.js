import React from "react";
import { render } from 'react-testing-library'
import GlobalState from '../contexts/GlobalState';

const AllTheProviders = ({ children }) => {
  return (
    <GlobalState>
        {children}
    </GlobalState>
  )
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from 'react-testing-library'

// override render method
export { customRender as render }