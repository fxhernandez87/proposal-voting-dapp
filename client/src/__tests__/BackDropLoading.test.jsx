import React from 'react';
import { render, fireEvent, waitForElement, cleanup } from 'test-utils';
import 'jest-dom/extend-expect';
import BackDropLoading from '../components/BackDropLoading';

beforeEach(cleanup);

test('Should create an overlay that says Initializing App', async () => {
  // Arrange
  const active = true;
  const children = <div />;
  const initializing = true;
  const error = false;

  const {queryByText, getByText, asFragment} = render(
    <BackDropLoading
      active={active}
      children={children}
      initializing={initializing}
      error={error}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
  expect(getByText(/Initializing App/i)).not.toBeNull();
  expect(queryByText(/_loading-overlay-transition-exit-active/i)).toBeNull();
});

test('Should create an overlay that says Fetching User', async () => {
  // Arrange
  const active = true;
  const children = <div />;
  const initializing = false;
  const error = false;

  const {queryByText, getByText, asFragment} = render(
    <BackDropLoading
      active={active}
      children={children}
      initializing={initializing}
      error={error}
    />,
  );

  expect(asFragment()).toMatchSnapshot();
  expect(queryByText(/_loading-overlay-transition-exit-active/i)).toBeNull();
  expect(getByText(/Fetching User/i)).not.toBeNull();
});


test('Should not show an overlay', async () => {
  // Arrange
  const active = false;
  const children = <div />;
  const initializing = false;
  const error = false;

  const {queryByText, asFragment} = render(
    <BackDropLoading
      active={active}
      children={children}
      initializing={initializing}
      error={error}
    />,
  );

  expect(asFragment()).toMatchSnapshot();
  expect(queryByText(/_loading_overlay_wrapper--active/i)).toBeNull();
});



test('Should fade overlay out on click', async () => {
  // Arrange
  const active = true;
  const children = <div />;
  const initializing = false;
  const error = false;

  const {getByText, getByTestId, asFragment} = render(
    <BackDropLoading
      active={active}
      children={children}
      initializing={initializing}
      error={error}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
  expect(getByText(/Fetching User/i)).not.toBeNull();
  await waitForElement(() =>
    getByTestId('wrapper')
  );

  fireEvent.click(getByText('Fetching User'));

  expect(asFragment()).toMatchSnapshot();
  expect(getByTestId('overlay')).toHaveClass('_loading-overlay-transition-exit-active');
});

test('Should show an error without spinner', async () => {
  // Arrange
  const active = false;
  const children = <div />;
  const initializing = false;
  const error = new Error('Test Error');

  const {getByText, queryByText, asFragment} = render(
    <BackDropLoading
      active={active}
      children={children}
      initializing={initializing}
      error={error}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
  expect(getByText(/Test Error/i)).not.toBeNull();
  expect(queryByText(/_loading_overlay_spinner/i)).toBeNull();
});

