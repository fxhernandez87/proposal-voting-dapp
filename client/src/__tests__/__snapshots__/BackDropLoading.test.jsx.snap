// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create an overlay that says Fetching User 1`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper _loading_overlay_wrapper--active css-ft4slc"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-df17o1"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        <div
          class="_loading_overlay_spinner css-50etie"
        >
          <svg
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </div>
        Fetching User
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Should create an overlay that says Initializing App 1`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper _loading_overlay_wrapper--active css-ft4slc"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-df17o1"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        <div
          class="_loading_overlay_spinner css-50etie"
        >
          <svg
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </div>
        Initializing App
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Should fade overlay out on click 1`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper _loading_overlay_wrapper--active css-ft4slc"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-df17o1"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        <div
          class="_loading_overlay_spinner css-50etie"
        >
          <svg
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </div>
        Fetching User
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Should fade overlay out on click 2`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper css-79elbk"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-1r098pj _loading-overlay-transition-exit-active"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        <div
          class="_loading_overlay_spinner css-50etie"
        >
          <svg
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </div>
        Fetching User
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Should not show an overlay 1`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper css-79elbk"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-1r098pj _loading-overlay-transition-exit-active"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        <div
          class="_loading_overlay_spinner css-50etie"
        >
          <svg
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
              stroke-miterlimit="10"
              stroke-width="2"
            />
          </svg>
        </div>
        Fetching User
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Should show an error without spinner 1`] = `
<DocumentFragment>
  <div
    class="_loading_overlay_wrapper _loading_overlay_wrapper--active css-ft4slc"
    data-testid="wrapper"
  >
    <div
      class="_loading_overlay_overlay css-df17o1"
      data-testid="overlay"
    >
      <div
        class="_loading_overlay_content css-42igfv"
      >
        Test Error
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
